File: LeetCode/cpp/287.cpp
class Solution {
public:
    int findDuplicate(vector<int> &nums) {
        auto size = nums.size();
        if (size == 0)
            return 0;
        int fast = nums[0], slow = nums[0];
        do {
            fast = nums[nums[fast]];
            slow = nums[slow];
        } while (fast != slow);
        fast = nums[0];
        while (fast != slow) {
            slow = nums[slow];
            fast = nums[fast];
        }
        return slow;
    }
};

File: LeetCode/python3/287.py
class Solution:
    def findDuplicate(self, nums: List[int]) -> int:
        p1, p2 = nums[0], nums[nums[0]]
        while nums[p1] != nums[p2]:
            p1 = nums[p1]
            p2 = nums[nums[p2]]
        p2 = 0
        while nums[p1] != nums[p2]:
            p1 = nums[p1]
            p2 = nums[p2]
        return nums[p1]
